### ISSUE TITLE
[question] Expand snippet after selection

### ISSUE THREAD

**Issue by balta2ar** [SirVer/ultisnips#682]
Hello. First of all, thank you for this awesome plugin! I'm new to ultisnips and I have little vimL experience so I decided to ask help on implementation here.

Say, I have a file with completion candidates located at `/tmp/candidates.txt` as follows:

```
338175 apple | Please update apple config https://www.google.com/config
364299 google | C++ coding style https://www.google.com/coding-style
337871 google | Implement simulation https://www.google.com/simulation
```

Whenever I type `RT` (any filetype) I want completion menu to show me candidates from that file. After `RT` I can input any of those lines partially. The candidates are filtered using fuzzy matching. When best candidate is expanded (selected), I get `RT<number>`.

For example, if I type `RT`, I want 5 top candidates to be shown in a completion menu.
After I typed `RT33`, I want only two lines to be show:

```
338175 apple | Please update apple config https://www.google.com/config
337871 google | Implement simulation https://www.google.com/simulation
```

If I press `C-y` (expand the snippet) now, the input should expand to `RT338175`.

After I typed `RTgosim` I want only one line to be shown:

```
337871 google | Implement simulation https://www.google.com/simulation
```

If I press `C-y` now, the input should expand to `RT337871`.

If there are many candidates to choose from, I want to be able to scroll over them (using `C-n`, `C-p`) and expand any candidate into usual form `RT<number>`.

Could you give me any hints on how to do that in vim? I thought ultisnips could be of a great help here and I could base my implementation on it.

So far I've tried the example from the docs:

```
snippet "RT:(\S+)" "RequestTracker" br
RT${1:`!p
snip.rv = match.group(1) if match.group(1) is not None else "something"`}
$0
endsnippet
```

However, I'm not sure how to display filtered candidates in a menu at this point.

Is it a good idea at all to use ultisnips for such a task? Maybe I need to explore YouCompleteMe or neocomplete or even `completefunc`?

Any help is appreciated. Thank you!

PS: I'm using YouCompleteMe.

**Comment by seletskiy** [comment]
@balta2ar: See `:h complete` in vim, it can be used to create custom completion menu. Keep that thread updated, I'm very curious about how exactly you will implement it!

**Comment by balta2ar** [comment]
Thanks for pointing out `:h complete`! I'll give it a read!

So far I've only managed to make a very sloppy, inefficient and hacky solution based on YouCompleteMe reference completer. Here is the code if you're interested: https://gist.github.com/balta2ar/7955094dbf88c4cf0a386f18d03ef447

Unfortunately, there is a slight problem with that. It only triggers on `RT` while I want it to also trigger on `RT:`. I'm going to ask a question about that in YouCompleteMe repository when I have more time.

On the other hand I've seen YouCompleteMe authors saying that it's not prose completer. Maybe there are other better ways to do that (maybe I need to read about vim's `omnicomplete`).

**Comment by seletskiy** [comment]
Seems to me that you're little bit overcomplicating solution. It should be as easy as calling complete function with array of completion args. 

By the way, it's not easy piece of the cake for making YCM works with : symbol, see how it's solved in https://github.com/kovetskiy/ycm-sh?files=1

**Comment by balta2ar** [comment]
> Seems to me that you're little bit overcomplicating solution. It should be as easy as calling complete function with array of completion args.

Maybe. I'm not very familiar with vim completion best practices.

In this case, how do I

1. display long candidates (including 123456 number, text and the link)
2. filter over this list of long candidates using the pattern that the user has typed so far after "RT:"
3. but when selection is made, insert only RT:123456, not the whole long candidate?

This is the issue I'm currently fighting with.

**Comment by seletskiy** [comment]
Can you describe the issue at top level? Maybe there is better solution. In my practice there were no issues with completing identifiers so entangled.

**Comment by balta2ar** [comment]
At top level the issue is as follows. At work we use RequestTracker system to track issues (tickets). We use ticket numbers in this format RT:123456 (or RT123456) a lot: in timesheet files, in commit messages, in the code as a reference to certain decisions. Often I remember ticket title (short description) but I don't remember the ticket number, of course.

For this reason I implemented a completer that would insert ticket numbers for me. Periodically I scan browser's tab titles for ticket numbers, retrieve ticket titles by those numbers and save a number, a title and a URL (candidates) into a file. Completion system uses that file to filter out tickets by a pattern and insert a number. I found it to be quite useful in my workflow.

**Comment by seletskiy** [comment]
@balta2ar: yep, I see. So, you can achieve that by calling `complete()`, but it will be not possible to use fuzzy matching, like YCM does. That's why there is YCM in first place. So, did you resolve that issue by using custom YCM completer?

> For this reason I implemented a completer that would insert ticket numbers for me. Periodically I scan browser's tab titles for ticket numbers, retrieve ticket titles by those numbers and save a number, a title and a URL (candidates) into a file. Completion system uses that file to filter out tickets by a pattern and insert a number. I found it to be quite useful in my workflow.

That is awesome! I like that kind of things. Keep going :+1:

**Comment by balta2ar** [comment]
Thanks! In fact, up to a recent moment I've been using YCM completer that I presented above. Unfortunately, it wasn't perfect: `RT:` (with a colon) didn't work, the completer didn't activate in all filetypes.

Recently I decided to give neovim a try. I also, for the sake of experiment, switched from YCM to [deoplete](https://github.com/Shougo/deoplete.nvim) to see whether I can feel the difference. And finally on these days I've implemented my first sketch of deoplete source: https://gist.github.com/balta2ar/f452af034da2d4027ae36c92d8cf9ba9

I have to say writing completion source for deoplete felt easier, and deoplete's docs are more intuitive. It was very easy to support both `RT` and `RT:` as a trigger. On top of that, it works in all filetypes.

I haven't spent much time using this new completer in an actual work yet so I'll see how it goes next week. Thank you for your interest and sorry for flooding this repo with comments not really related to ultisnips! :)
